alf/algorithms/generator.py

```diff
@@ -395,69 +550,203 @@ def _svgd_grad(self, inputs, outputs, loss_func, entropy_regularization):
+        else:
+            aug_outputs = outputs  # [N, D']
+            aug_outputs2 = outputs2  # [N2, D']
         loss_inputs = outputs2
         loss = loss_func(loss_inputs)
         if isinstance(loss, tuple):
             neglogp = loss.loss
         else:
             neglogp = loss
-        loss_grad = torch.autograd.grad(neglogp.sum(), outputs2)[0]  # [N2, D]
+        loss_grad = torch.autograd.grad(neglogp.sum(),
+                                        loss_inputs)[0]  # [N2, D]
 
-        # [N2, N], [N2, N, D]
-        kernel_weight, kernel_grad = self._rbf_func2(outputs2, outputs)
+        # [N2, N], [N2, N, D']
+        kernel_weight, kernel_grad = self._rbf_func2(aug_outputs2.detach(),
+                                                     aug_outputs.detach())
         kernel_logp = torch.matmul(kernel_weight.t(),
                                    loss_grad) / num_particles  # [N, D]
-        grad = kernel_logp - entropy_regularization * kernel_grad.mean(0)
-        loss_propagated = torch.sum(grad.detach() * outputs, dim=-1)
+
+        loss_prop_kernel_logp = torch.sum(
+            kernel_logp.detach() * outputs, dim=-1)
+        loss_prop_kernel_grad = torch.sum(
+            -entropy_regularization * kernel_grad.mean(0).detach() *
+            aug_outputs,
+            dim=-1)
+        loss_propagated = loss_prop_kernel_logp + loss_prop_kernel_grad
 
         return loss, loss_propagated
 
-    def _gfsf_grad(self, inputs, outputs, loss_func, entropy_regularization):
+    def _gfsf_grad(self,
+                   inputs,
+                   outputs,
+                   loss_func,
+                   entropy_regularization,
+                   transform_func=None):
         """Compute particle gradients via GFSF (Stein estimator). """
         assert inputs is None, '"gfsf" does not support conditional generator'
+        if transform_func is not None:
+            outputs, extra_outputs, _ = transform_func(outputs)
+            aug_outputs = torch.cat([outputs, extra_outputs], dim=-1)
+        else:
+            aug_outputs = outputs
+        score_inputs = aug_outputs.detach()
         loss_inputs = outputs
         loss = loss_func(loss_inputs)
         if isinstance(loss, tuple):
             neglogp = loss.loss
         else:
             neglogp = loss
-        loss_grad = torch.autograd.grad(neglogp.sum(), outputs)[0]  # [N2, D]
+        loss_grad = torch.autograd.grad(neglogp.sum(),
+                                        loss_inputs)[0]  # [N2, D]
+        logq_grad = self._score_func(score_inputs) * entropy_regularization
 
-        logq_grad = self._score_func(outputs)
-        grad = loss_grad - entropy_regularization * logq_grad
-        loss_propagated = torch.sum(grad.detach() * outputs, dim=-1)
+        loss_prop_neglogp = torch.sum(loss_grad.detach() * outputs, dim=-1)
+        loss_prop_logq = torch.sum(logq_grad.detach() * aug_outputs, dim=-1)
+        loss_propagated = loss_prop_neglogp + loss_prop_logq
```

Comment:
according to line 459 of the old file, this should be `loss_prop_neglogp - loss_prop_logq`

Reply:
I added a negation in the return of self._score_func (Line 518) for the purpose of exactly matching the formula of \nabla\log q.